### ISSUE TITLE
PR/Branch Up-To-Date Checker - Support default/mainline branches other than 'master'

### ISSUE THREAD

**Issue by shadeon** [kaxil/boring-cyborg#7]
We've just enabled the bot (nice work by the way, very helpful 👍 ) in our repository, but we aren't getting checks noting if the PR is up to date with the target branch or not.

Having looked at the code am I right in saying it's hard-coded to only work against "master" specifically? This is notable because "master" is not our default/mainline branch.

If the above is the case, would it be possible in addition to check the mainline to have a second check for the _target_ branch?

Would it also be possible to list the number of commits behind master/target branch (appreciate that this number may become stale if master moves on further).

Thanks!

**Comment by kaxil** [comment]
>Having looked at the code am I right in saying it's hard-coded to only work against "master" specifically? This is notable because "master" is not our default/mainline branch.

That is correct.

>If the above is the case, would it be possible in addition to check the mainline to have a second check for the target branch?

Yes, I will expose a variable to define a target branch.

> Would it also be possible to list the number of commits behind master/target branch (appreciate that this number may become stale if master moves on further).

I feel it might not be that helpful if the number becomes stale. Would love to know your use-case

**Comment by shadeon** [comment]
>Yes, I will expose a variable to define a target branch.

Thanks!

Our use case largely comes down to our branching (fairly typical model, I think) and just trying to keep developers informed.  Our structure looks a bit like this:

```
master
    |  Feature X
        | work branch
        | work branch
    |  Feature Y 
    | Bug work branch
```

Features are built by multiple developers on work branches and merging into a feature branch. When the feature branch has met the definition of done, it is merged into master. This is all done with PRs (work to feature, and feature to master). Small bug fixes not related to a feature are PR'ed directly into master. Any changes in master are synced downstream shortly after merging.

We've just migrated to Github from another git provider who shows in PRs how many commits behind your target branch you are. AFAIK to get that information in Github you have to go and configure a branch comparison, outside of the PR.

Ideally we'd like a check on our PRs (we've configured the existing check not to block) that gives developers a sense of how out of date they are both with master, but __also__ the feature branch they are targeting, which may or may not warrant a sync before merge and re-review. I'm a bit concerned that if the check flags for both a single commit small bugfix and also a much larger feature merge, they will stop bothering to go check.

A better alternative could be to construct the branch comparison link and provide that as a details link in the check (e.g `<repo>/compare/work/ml/1234...master`) so it's easy to view the results.

So to summarise:

1. One check that compares the branch to master (which would require a configurable name in our case, since it's not named master in our case)
2. A second separate check that compares to the target branch
3. Put a details link in both checks that sends you to the Git branch comparison page for the work<-master and work<-feature syncs.

Point 1 is necessary for us to use this feature of the bot at all, but 2 and 3 would be amazing additions for us.

Thanks again!

**Comment by kaxil** [comment]
> > Yes, I will expose a variable to define a target branch.
> 
> Thanks!
> 
> Our use case largely comes down to our branching (fairly typical model, I think) and just trying to keep developers informed. Our structure looks a bit like this:
> 
> ```
> master
>     |  Feature X
>         | work branch
>         | work branch
>     |  Feature Y 
>     | Bug work branch
> ```
> 
> Features are built by multiple developers on work branches and merging into a feature branch. When the feature branch has met the definition of done, it is merged into master. This is all done with PRs (work to feature, and feature to master). Small bug fixes not related to a feature are PR'ed directly into master. Any changes in master are synced downstream shortly after merging.
> 
> We've just migrated to Github from another git provider who shows in PRs how many commits behind your target branch you are. AFAIK to get that information in Github you have to go and configure a branch comparison, outside of the PR.
> 
> Ideally we'd like a check on our PRs (we've configured the existing check not to block) that gives developers a sense of how out of date they are both with master, but **also** the feature branch they are targeting, which may or may not warrant a sync before merge and re-review. I'm a bit concerned that if the check flags for both a single commit small bugfix and also a much larger feature merge, they will stop bothering to go check.
> 
> A better alternative could be to construct the branch comparison link and provide that as a details link in the check (e.g `<repo>/compare/work/ml/1234...master`) so it's easy to view the results.
> 
> So to summarise:
> 
> 1. One check that compares the branch to master (which would require a configurable name in our case, since it's not named master in our case)
> 2. A second separate check that compares to the target branch
> 3. Put a details link in both checks that sends you to the Git branch comparison page for the work<-master and work<-feature syncs.
> 
> Point 1 is necessary for us to use this feature of the bot at all, but 2 and 3 would be amazing additions for us.
> 
> Thanks again!

Thanks for the detailed explanation. Point (1) makes sense in general and I would try to get it done this week. Also (2) and (3) sound plausible too, just need to find time for them.

I am open to contributions too incase you want to, if not I will try to get this done over the weekend.

**Comment by shadeon** [comment]
>I am open to contributions too incase you want to, if not I will try to get this done over the weekend.

Awesome, I'm happy to take on 2 & 3 since they are fairly specific requests, though since it will take me a while to do so if you can implement the configurable default branch name in the meantime that would be great 😄

**Comment by shadeon** [comment]
Thanks for implementing this - put the new config setting in this morning, but still not seeing the check appear (positive or negative). Is there a way of seeing some kind of log/troubleshooting this, or have the changes just not been made live yet and I'm being impatient 😄

**Comment by kaxil** [comment]
> Thanks for implementing this - put the new config setting in this morning, but still not seeing the check appear (positive or negative). Is there a way of seeing some kind of log/troubleshooting this, or have the changes just not been made live yet and I'm being impatient 😄

Interestingly, the code is live and I tested it at https://github.com/kaxil/test

The main branch over there is: **develop**
PR I created against **develop**: https://github.com/kaxil/test/pull/3

![image](https://user-images.githubusercontent.com/8811558/76210821-cef46f00-61fc-11ea-8f7d-dccb76942c87.png)

I feel the problem might be that the bot doesn't have read permissions on your repo's files if your repo is private. Currently, I had requested minimal permissions (read permissions on a single file: `.github/boring-cyborg.yml`

Can you test it against a public repo too?

Secondly, if it doesn't work, do you think you and other users might be comfortable giving read access on all files of your private repo?

**Comment by shadeon** [comment]
> I feel the problem might be that the bot doesn't have read permissions on your repo's files if your repo is private. Currently, I had requested minimal permissions (read permissions on a single file: `.github/boring-cyborg.yml`

I suspect you are correct, although notably the automatic label functionality works fine so guessing it may be another request that is being blocked instead (list of branches, or commits?).

> Can you test it against a public repo too?

Will do.

> Secondly, if it doesn't work, do you think you and other users might be comfortable giving read access on all files of your private repo?

I can't speak for other users, but I know for our organisation giving read access for all files in the repo would be a no-go.

**Comment by kaxil** [comment]
>I suspect you are correct, although notably the automatic label functionality works fine so guessing it may be another request that is being blocked instead (list of branches, or commits?).

With **automatic label** (labeler) it just checks the filenames in the PR and hence it does not need read permission across all the files in the private repo. So it works fine.

>I can't speak for other users, but I know for our organisation giving read access for all files in the repo would be a no-go.

I think it would be the case with everyone, or else why would you want to keep the repo private 😄 .
So, in that case, the solution would be to fork and deploy the bot in your environment and give it permissions in it. The only downside is you will have to either maintain the fork or sync the code from the master branch here.

**Comment by shadeon** [comment]
> With **automatic label** (labeler) it just checks the filenames in the PR and hence it does not need read permission across all the files in the private repo. So it works fine.

After looking at the permissions you can grant github apps I see what you mean - it's not granular enough for you to grant branches and commits _without_ blobs.

We'll fork it and run ourselves.  I'll make sure any improvements we make get raised as PR's back upstream 😄  Thanks for your help!

**Comment by kaxil** [comment]
> > With **automatic label** (labeler) it just checks the filenames in the PR and hence it does not need read permission across all the files in the private repo. So it works fine.
> 
> After looking at the permissions you can grant github apps I see what you mean - it's not granular enough for you to grant branches and commits _without_ blobs.
> 
> We'll fork it and run ourselves. I'll make sure any improvements we make get raised as PR's back upstream 😄 Thanks for your help!

Awesome, appreciate it. Looking forward to your contributions :)
